### ISSUE TITLE
A2CのSampleLayerのノイズについて

### ISSUE THREAD

**Issue by tatsuya-ogawa** [icoxfog417/baby-steps-of-rl-ja#20]
### 指摘事項
SampleLayerのノイズの計算で下記の式となっていますが、
これだとノイズが大きいと思ったのですが、どのような意図でこの式となっているのでしょうか?
`return tf.argmax(x - tf.log(-tf.log(noise)), axis=1)`

### 指摘箇所
* [ ] Day4: 強化学習に対するニューラルネットワークの適用

ページ番号: p
code4-24

**Comment by icoxfog417** [comment]
この実装はOpenAI baselineのA2Cの実装を参考にしています。

OpenAI baselineの実装を以下のようにたどると、該当の処理にたどり着くことができます。ちょっと長くなりますがご容赦ください。

1. [構築したA2Cのモデルを使い、`build_policy`によりpolicyを作成](https://github.com/openai/baselines/blob/master/baselines/a2c/a2c.py#L188)
2. [`build_policy`の返却値(policyの実体)は`PolicyWithValue`](https://github.com/openai/baselines/blob/master/baselines/common/policies.py#L168)
3. [`PolicyWithValue`において、actionは`self.pd`からsampleされる](https://github.com/openai/baselines/blob/master/baselines/common/policies.py#L52)
4. [`self.pd`は環境におけるactionの種別から選択される(`make_pdtype`)](https://github.com/openai/baselines/blob/master/baselines/common/policies.py#L47)
5. [Catcherの場合は離散値のactionなので`CategoricalPdType`となる](https://github.com/openai/baselines/blob/master/baselines/common/distributions.py#L284)
6. [`CategoricalPdType`における`pd`は`CategoricalPd`となる(親クラスの`pdfromflat`の挙動を継承)](https://github.com/openai/baselines/blob/master/baselines/common/distributions.py#L66)
7. [`CategoricalPd`における`sample`の実装は以下となる](https://github.com/openai/baselines/blob/master/baselines/common/distributions.py#L199)

```py
def sample(self):
    u = tf.random_uniform(tf.shape(self.logits), dtype=self.logits.dtype)
    return tf.argmax(self.logits - tf.log(-tf.log(u)), axis=-1)
```

**Comment by tatsuya-ogawa** [comment]
@icoxfog417 
有難うございます。
OpenAI Baselineのリポジトリでも類似の質問されている方がいらっしゃったので、参考としてリンク貼らせて頂きます。
[OpenAI Baselineのissue](https://github.com/openai/baselines/issues/172)
ガンベル分布というものを使用されているということですね。
了解しました。

**Comment by icoxfog417** [comment]
詳細まで調べて頂きありがとうございます！どんな分布か気にしていなかったため、この点について次版で記載しておこうと思います。

**Comment by tatsuya-ogawa** [comment]
Gumbel分布について、この方のブログが勉強になったので、リンク貼らせて頂きます。
[Gumbel Max Trickについて勉強した](http://peluigi.hatenablog.com/entry/2018/06/21/142753)
同じ疑問でissue検索された方のご参考になればと思います。

**Comment by icoxfog417** [comment]
こちらも参考になりました

[Categorical Reparameterization with Gumbel-Softmax ](http://musyoku.github.io/2016/11/12/Categorical-Reparameterization-with-Gumbel-Softmax/)
